templates/shift-scheduling/main.go

```diff
@@ -1,370 +1,248 @@
+				}
+			}
+		}
+	}
 
-	// We define some helper variables that are useful later.
-	nShifts := input.Days * 3
-	workers := model.NewDomain(model.NewRange(1, input.Workers))
-	typeMap := [3]shiftType{"morning", "day", "night"}
-	prefShiftType := map[int]shiftType{}
-	for _, v := range input.Preferences.ShiftType {
-		prefShiftType[int(v.Worker)] = v.Type
+	// initialize demand ids
+	for i, demand := range input.RequiredWorkers {
+		demand.RequiredWorkerID = i
+		input.RequiredWorkers[i] = demand
 	}
 
-	// We track the following variables:
-	// shifts holds one domain for each shift (3 per day in order morning, day,
-	// night) initially it holds all workers available.
-	shifts := store.Repeat(schedule, nShifts, workers)
-
-	// The number of workers active in a schedule.
-	workerCount := store.NewVar(schedule, 0)
-
-	// We also want to maximize worker happiness by fulfilling as many
-	// preferences as possible.
-	happiness := store.NewVar(schedule, 0)
-
-	// As a first step, we can now remove the workers that are unavailable for
-	// certain full days. This results as a new (partial) schedule.
-	changes := []store.Change{}
-
-	for _, v := range input.Unavailable {
-		for _, d := range v.Days {
-			morning := (d - 1) * 3
-			day := morning + 1
-			night := day + 1
-			changes = append(changes,
-				shifts.Remove(morning, []int{int(v.Worker)}),
-				shifts.Remove(day, []int{int(v.Worker)}),
-				shifts.Remove(night, []int{int(v.Worker)}),
-			)
+	demandCovering := map[int][]Assignment{}
+	for _, demand := range input.RequiredWorkers {
+		demandCovering[demand.RequiredWorkerID] = []Assignment{}
+		for i, potentialAssignment := range potentialAssignments {
+			if (potentialAssignment.Start.Before(demand.Start.Time) || potentialAssignment.Start.Equal(demand.Start.Time)) &&
+				(potentialAssignment.End.After(demand.End.Time) || potentialAssignment.End.Equal(demand.End.Time)) {
+				potentialAssignments[i].DemandsCovered = append(potentialAssignments[i].DemandsCovered, demand)
+				demandCovering[demand.RequiredWorkerID] = append(demandCovering[demand.RequiredWorkerID], potentialAssignment)
+			}
 		}
 	}
-	schedule = schedule.Apply(changes...)
 
-	// Now we define our objective value. This is the quantity we try to
-	// minimize. It is the weighted sum of workers and happiness. Happiness is
-	// negative as we want to maximize it while minimizing the worker count.
-	workerCountImportance := 10
-	schedule = schedule.Value(func(s store.Store) int {
-		return workerCountImportance*workerCount.Get(s) - happiness.Get(s)
-	})
-
-	// Now we define our constraint containing a propagator. This will remove
-	// workers from domains based upon the workers in other domains.
-	constraint := newBreakConstraint(shifts, 2)
-
-	// We invoke the propagator to propagate the initial values.
-	schedule.Propagate(constraint.propagate)
-
-	if constraint.isProvenInfeasible(schedule) {
-		return nil, fmt.Errorf("input can not be solved")
+	m := mip.NewModel()
```

Comment:
Like mentioned above let's extract the mip model building into a private helper function